### PR TITLE
Add Akismet filtering of doc content.

### PR DESCRIPTION
**Pull request by dcavins**
If Akismet is configured for the site, send the content over to Akismet for inspection. If Akismet responds that the content looks like spam, set the post status to a new status, "bp_docs_pending," so that site admins can choose to either publish or delete the new content. Docs that are pending are visible to the creator and site admins only.

### CODE REVIEW

**Comment by dcavins** [comment]
A little more information about how this changes the various directories: If a doc is identified as spam by Akismet, it is given a `post_status` of `bp_docs_pending` which means that the doc will require moderation in order to be made public. In the Docs admin view, a pending doc shows up like this:
docs-admin

In the main directory, if you're a site admin, a pending doc shows up like this:
docs-directory-as-admin

If you're not a site admin, pending docs are not included in the directory view:
docs-directory-non-admin
 
If you're the creator of the doc that was marked for moderation, you can still see and edit the doc, but you can't publish it.
doc-creator-can-see-moderated-docs

Another useful side effect of this change is that a one-liner filter could be used to add a moderation workflow for all docs (marking all new, non-admin-authored docs as `bp_docs_pending`), if that's appropriate for the site.

File: includes/addon-akismet.php [hunk 1]
```diff
@@ -0,0 +1,185 @@
...
+	 */
+	public function check_for_spam( $save_args, $bdq_object, $passed_args ) {
+		// Build data package for Akismet.
+		$akismet_package = $this->build_akismet_data_package( $save_args, $passed_args );
+
+		// Check with Akismet to see if this is spam.
+		$akismet_response = $this->send_akismet_request( $akismet_package, 'check', 'spam' );
+
+		// Spam.
+		if ( 'true' == $akismet_response['bp_docs_as_result'] ) {
```
Comment (boonebgorges) [inline includes/addon-akismet.php]: It really sends back the string `'true'`?

File: includes/addon-akismet.php [hunk 2]
```diff
@@ -0,0 +1,185 @@
...
+			/**
+			 * Fires after a doc has been proven to be spam, but before officially being marked as spam.
+			 *
+			 * @since 2.0.0
```
Comment (boonebgorges) [inline includes/addon-akismet.php]: 2.1.0

File: includes/addon-moderation.php [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+	/**
+	 * Constructor.
+	 *
+	 * @since 2.1.0
+	 */
+	public function __construct() {
+	}
+
+	/**
+	 * Hook the Folders functionality into Docs.
```
Comment (boonebgorges) [inline includes/addon-moderation.php]: copypasta

File: includes/addon-moderation.php [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+	 *
+	 * @since 2.1.0
+	 */
+	public function add_hooks() {
+		// Register a custom status that's a lot like the built-in "pending" status.
```
Comment (boonebgorges) [inline includes/addon-moderation.php]: Why at `9`? I assume this was necessary for some reason.

File: includes/addon-moderation.php [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+
+	/**
+	 * Register custom status for pending BP Docs.
+	 *
+	 * @since 2.1.0
+	 */
+	public function register_docs_pending_status() {
+		$args = array(
+			'label'                     => _x( 'Awaiting Moderation', 'Status General Name', 'buddypress-docs' ),
+			'label_count'               => _n_noop( 'Awaiting Moderation (%s)',  'Pending (%s)', 'buddypress-docs' ),
```
Comment (boonebgorges) [inline includes/addon-moderation.php]: `_n_noop()` is for registering singular + plural - why are the singular and plural different?

File: includes/component.php [hunk 6]
```diff
@@ -767,7 +767,7 @@ public function delete_doc_activity( $new_status, $old_status, $post ) {
 	 * @return str $link The filtered permalink
 	 */
 	function filter_permalinks( $link, $post, $leavename, $sample ) {
-		if ( bp_docs_get_post_type_name() == $post->post_type ) {
+		if ( bp_docs_get_post_type_name() == $post->post_type && ! empty( $post->post_name ) ) {
```
Comment (boonebgorges) [inline includes/component.php]: Why this change? Don't pending items have `post_name`?

File: includes/query-builder.php [hunk 7]
```diff
@@ -541,8 +541,24 @@ function save( $passed_args = array() ) {
...
+			 * Fires before the doc has been saved.
+			 *
+			 * @since 2.1.0
+			 *
+			 * @param array  $r    The parameters to be used in wp_insert_post().
+			 * @param object $this The BP_Docs_Query object.
+			 * @param array  $args The passed and filtered parameters for the doc
+			 *                     about to be saved.
+			 */
+			$r = apply_filters( 'bp_docs_args_before_save', $r, $this, $args );
```
Comment (boonebgorges) [inline includes/query-builder.php]: maybe better to clarify `bp_docs_post_args_before_save` or something like that?

File: includes/component.php (hunk 6 above)
Comment (dcavins) [reply includes/component.php]: This was necessary when I was using the WP built-in pending status, which doesn't allow a `post_name` to be generated. There were other problems, too, so I ended up using a custom status (that was your suggestion, and it was the right move--way easier than handling all the odd behavior of the built-in pending status).

File: includes/addon-moderation.php (hunk 4 above)
Comment (dcavins) [reply includes/addon-moderation.php]: It's been too long for me to remember why this was necessary. I'll remove it and see what happens. :)

File: includes/addon-akismet.php (hunk 1 above)
Comment (dcavins) [reply includes/addon-akismet.php]: Ha ha ha, no, thankfully. 💯

File: includes/addon-moderation.php (hunk 5 above)
Comment (dcavins) [reply includes/addon-moderation.php]: It seems that `register_post_status()` or rather the global `$wp_post_statuses` expects the label_count parameter to be an `_n_noop()`. They shouldn't be different, though, that's just me being not very smart, ha ha.

https://developer.wordpress.org/reference/functions/register_post_status/

File: includes/addon-akismet.php (hunk 1 above)
Comment (dcavins) [reply includes/addon-akismet.php]: OK, well then I take it back:
```
akismet_response: array(2) {
  [0]=>
  object(Requests_Utility_CaseInsensitiveDictionary)#1403 (1) {
    ["data":protected]=>
    array(5) {
      ["server"]=>
      string(5) "nginx"
      ["date"]=>
      string(29) "Fri, 22 Dec 2017 15:58:16 GMT"
      ["content-type"]=>
      string(25) "text/plain; charset=utf-8"
      ["content-length"]=>
      string(1) "4"
      ["x-akismet-guid"]=>
      string(32) "long-string-xxxxx"
    }
  }
  [1]=>
  string(4) "true"
}
```
I feel a little better, anyway.

**Comment by dcavins** [comment]
Thanks for the review Boone!

File: includes/addon-akismet.php (hunk 1 above)
Comment (boonebgorges) [reply includes/addon-akismet.php]: I had a feeling they probably did, but I wanted to be sure in case you were getting fooled by the loose `==` :-D
